### DISCUSSION TITLE
how to port a minizinc model into Z3 using C# API via PowerShell

### DISCUSSION THREAD

**Discussion by eosfor** [Z3Prover/z3#5790] [Q&A] [answered]
Hello colleagues,

I have a minizinc model, which is supposed to solve an "assignment" problem. Basically, having a list of Virtual Machine objects with their attributes, I'm trying to map them to their counterparts in Azure, in such a way that the counterpart has enough vCPUs and RAM. Additionally, I'm trying to minimize the target cost and maximize the performance attributes.

To make it a bit more interactive, easier to use from the command line I'm want to use .NET interface through PowerShell. It allows to combine all necessary tools under the same umbrella, so one does not need to install anything except PowerShell, basically

The [minizinc model is here](https://github.com/eosfor/z3tests/blob/main/vmCostsCalculation-integer.mzn).
The [PowerShell script is here](https://github.com/eosfor/z3tests/blob/main/z3.ps1)
The data files and z3 binaries to make it work are also there.

So what I'm trying to do is literally map it to z3.

First, I'm preparing necessary data by reading a few CSV files and indexing that data to hashtables, just to speed up access

```powershell
# import raw VM data and convert strings to ints
$sourceVMs = Import-Csv .\vmdata.csv
$sourceVMs  | % { $_.cpu = [int]$_.cpu; $_.ram = [int]$_.ram; $_.datadisk = [int]$_.datadisk; }

$targetSizes = import-csv ".\vmCostACUData.csv"

## hash tables with data to simplify search
$sourceVmHt = @{}
$targetVmHt = @{}

$sourceVMs | % {
    $sourceVmHt[$_.vmID] = $_
}

$targetSizes | % {
    $targetVmHt[$_.Name] = $_
}
##
```

Next, I basically create the necessary enums, as I did in the minizinc model. They are supposed to limit possible values of the decision variables and serve as indexes in some cases.

```powershell
# enum existingVMs;
# enum vmSizes;

$existingVMs = $ctx.MkEnumSort("existingVMs", $sourceVMs.vmid)
$vmSizes = $ctx.MkEnumSort("vmSizes", $targetSizes.Name)
```

Create a bunch of decision variables, each of which is supposed to contain a name of that target VM size, calculated by the model, and store them in the array

```powershell
$selectedSize = 
    $sourceVMs | % {
        $ctx.MkConst($_.vmid, $vmSizes)
    }
```

and add two main constraints

```powershell
# constraint forall(vm in existingVMs)(
#    vmSizeCPU[selectedSize[vm]] >=  vmCPU[vm]
# );


## assuming $_.ToString() returns the name of the variable
## and $_ it's value
$selectedSize | % {
    $a = $ctx.MkNumeral($sourceVmHt[$_.ToString()].cpu, $ctx.MkIntSort())
    $b = $ctx.MkNumeral($targetVmHt[$_].vCPUs, $ctx.MkIntSort())
    $s.Add( $ctx.MkGt($a, $b))
}

# constraint forall(vm in existingVMs)(
#     vmSizeRAM[selectedSize[vm]] >= vmRAM[vm]
# );

$selectedSize | % {
    $a = $ctx.MkNumeral($sourceVmHt[$_.ToString()].ram, $ctx.MkIntSort())
    $b = $ctx.MkNumeral($targetVmHt[$_].MemoryGB, $ctx.MkIntSort())
    $s.Add( $ctx.MkGt($a, $b))
}
```

At this point I expect the model to return at least something, but it returns an empty model

```console
PS C:\Work\z3tests> c:\Work\z3tests\z3.ps1
SATISFIABLE
PS C:\Work\z3tests> $m

NumConsts  : 0
ConstDecls : {}
Consts     : {}
NumFuncs   : 0
FuncDecls  : {}
Decls      : {}
NumSorts   : 0
Sorts      : {}
```

At least this is what I think I'm doing :)

The question here is - what I'm doing wrong, and what should I do to make it right?
And then how to make the model to try optimize the objective functions?

**Comment by NikolajBjorner** [comment]
ctx.MkNumeral tries to parse your string as a numeric constant not as a variable. Use MkConst

**Reply by eosfor** [reply]
Thanks, @NikolajBjorner 

Let me probably put it a bit differently. To show where I am, I've put together a small project in C#. I hope this way it will be easier to explain. The project is [here](https://github.com/eosfor/z3testps), the actual piece of code is [here](https://github.com/eosfor/z3testps/blob/master/z3testps/StartZ3ModelCalculation.cs).

To my understanding, the main points are these. First I should create a set of "decision variables", the model should then decide their values. I assume that in Z3 they are called "constants", but I may be wrong :)/ So I create a set of such variables, and for simplicity store them in the dictionary.

The domain of each variable is limited by the enum, I derive from the source data

```csharp
            var vmSizes = ctx.MkEnumSort("vmSizes", TargetVMIndex.Keys.ToArray());

            var selectedSize = new Dictionary<string, Expr>();


            for (int i = 0; i < SourceVM.Length; i++)
            {
                var name = SourceVM[i].Properties["vmid"].ToString();
                selectedSize.Add(name, ctx.MkConst(name, vmSizes));
            }
```

So here I assume, that the model should decide the value for each variable, based on constraints I want to supply. Constraints go as follows. Here I'm using the enum values as indexes into the source data and into the dictionary of variables. But for them, I need to use the value of the variable as an index in another data dictionary.  And that is what I don't know how to do. The code below obviously does not work, and the question here is how to make it work? 
```csharp
            //constraint forall(vm in existingVMs)(
            //    vmSizeRAM[selectedSize[vm]] >= vmRAM[vm]
            //);
            foreach (var el in existingVMs.Consts)
            {
                var sourceObject = SourceVMIndex[el.ToString()];
                var targetObject = selectedSize[el.ToString()];
                s.Add(ctx.MkGt(TargetVMIndex[targetObject].vCPUs, sourceObject.cpu));
            }

            //constraint forall(vm in existingVMs)(
            //    vmSizeCPU[selectedSize[vm]] >= vmCPU[vm] * 0.8
            //);
            foreach (var el in existingVMs.Consts)
            {
                var sourceObject = SourceVMIndex[el.ToString()];
                var targetObject = selectedSize[el.ToString()];
                s.Add(ctx.MkGt(TargetVMIndex[targetObject].MemoryGB, sourceObject.ram));
            }

```

Perhaps I don't understand some concepts, as I'm trying to reflect my knowledge of minizinc into this and it may be not relevant. Would be really appreciated for any explanations or clues.

Thanks

**Reply by eosfor** [reply]
I spent some more time experimenting. Unfortunately, there are not that many "digestible" tutorials on C# APIs, so from time to time I'm just blindly trying this or that. But anyway, What I've done so far is [this](https://github.com/eosfor/z3testps/blob/master/z3testps/StartZ3ModelCalculation.cs). In general, it is a PowerShell command implemented in C#. It takes two pieces of data, the list of "source" VMs and the lit of "target" sizes. And it tries to match one into another by a few parameters, like CPU and RAM + tries to optimize two objective functions. The "source list" has 87 rows in it, and the "target" on has 240 rows. So when I run this with the whole input it never ends. The sample usage is [here](https://github.com/eosfor/z3testps/blob/master/build-test-commands.ps1)

The question here is, do I do something wrong and where?

As a side note, minizing+gecode solves the same problem in less than a second. and minizinc + or-tools does it in around 5 seconds. So I assume that Z3 should do this somewhere around those numbers. So I assume I do smth wrong here, but I'm stuck and have no idea where to look :(

**Comment by NikolajBjorner** [comment, marked as answer]
I don't expect z3 to do well on problem domains it is not tuned for. It is not built specifically for OR domains. 
These domains where z3 is not a suitable tool include LP problems, especially problems where constants are provided with a large decimal expansion and IP problems with bounded numbers. The sweet spot in some cases is discrete domains where you can leverage SMT features. See http://theory.stanford.edu/~nikolaj/supercharging.html for a case study.

**Reply by eosfor** [reply]
quick one, in this call `public ArraySort MkArraySort(Sort domain, Sort range)` what does `domain` and `range` mean?

**Reply by eosfor** [reply]
I've read this message a few times :) and I feel like I'm stupid. Seems like I need to back up and refresh (get) some (more) knowledge. Let me ask you this way. Do I do it correctly? Is it possible to re-write this model in such a way, that it works fast enough? any clues on what to do for that?

I also trying to digest the case study you shared :)

**Reply by NikolajBjorner** [reply]
For arrays, see https://smtlib.cs.uiowa.edu/theories-ArraysEx.shtml
The array sort constructor takes two arguments, one for the index set (domain) and the other the range of the array.

**Reply by eosfor** [reply]
Perhaps it is a problem of "translation" of the word "range" + some missing theoretical knowledge in my head :).  What I meant to ask, is, perhaps, how are `domain` and `range` related to each other? For example, if the `domain` is a set of all letters from "A" to "Z", what is the `range`, is it a subset of the domain like a set of letters from `F` to `L`? Or it could be something completely different like a set of integers? Should the `range` have the same cardinality as the `domain`. If we map this to the classic arrays in c# like `int[]a = new int[] {1,2,3};`, what is the `range` and what is the `domain` in this case?

I'm sorry for asking dumb questions :(, I'm just a poor systems engineer working mostly in planning azure infrastructures trying to look wider :D

**Reply by NikolajBjorner** [reply]
For one dimensional arrays in programming languages such as C, Java, C# the domains are typically 
machine integers, typically uint32 or non-negative int. Some other languages allow other kinds of domains.
The range of int[] a is the sort of machine integers "int".
The range of double[] is the sort of IEEE floats of some suitable size.

The "theory" of arrays is really a first-order formalism for applicative maps.
Proofs of programs that manipulate arrays can be carried out using the theory of arrays when creating verification conditions.

**Reply by eosfor** [reply]
Does it mean that arrays in Z3 is not for, kind of, "storing" some data and iterating through it, but rather for something different?

In this case, and this is the question I broke my mind against, say I have a "decision variable" that is supposed to serve as an index into some normal c# array: `var x1 = ctx.MkConst("i", ctx.IntSort)`, and the array, like `int[]a = new int[] {1,2,3, ...}` . How can I use this variable as an index into that array in C# API, and does this question even makes sense? :)

**Reply by eosfor** [reply]
I've done another attempt :). First, I added [the model](https://github.com/eosfor/z3testps/blob/master/z3testps/StartORToolsModelCalculation.cs) based on `or-tools`. What it does is just create a huge array (A) of `source x target` where each A[i, j] indicates if a particular source maps to a particular target.  And surprisingly, it works.

Then I decided to try to replicate the same approach to Z3, [here](https://github.com/eosfor/z3testps/blob/master/z3testps/StartZ3ModelCalculation.cs). And it did not work ;(. For some reason, it does not optimize objectives. It returns them like:

```console
## MAX objective
Lower         : oo
Upper         : oo
Value         : oo
LowerAsVector : {1, 0, 0}
UpperAsVector : {1, 0, 0}

## MIN objective
Lower         : (* (- 1) oo)
Upper         : oo
Value         : (* (- 1) oo)
LowerAsVector : {-1, 0, 0}
UpperAsVector : {1, 0, 0}
```

Also, it seems like the model has over 15k variables at the end. Is it good or bad or does not matter much?

```
**NumConsts  : 15788**
ConstDecls : {(declare-fun vmN80-Standard_M32ls () Int), (declare-fun vmN6-Standard_D32s_v3 () Int), (declare-fun vmN25-Standard_M128ms () Int), (declare-fun
             vmN45-Standard_GS1 () Int)…}
Consts     : {[(declare-fun vmN80-Standard_M32ls () Int), 0], [(declare-fun vmN6-Standard_D32s_v3 () Int), 0], [(declare-fun vmN25-Standard_M128ms () Int),
             0], [(declare-fun vmN45-Standard_GS1 () Int), 0]…}
NumFuncs   : 0
FuncDecls  : {}
Decls      : {(declare-fun vmN80-Standard_M32ls () Int), (declare-fun vmN6-Standard_D32s_v3 () Int), (declare-fun vmN25-Standard_M128ms () Int), (declare-fun
             vmN45-Standard_GS1 () Int)…}
NumSorts   : 0
Sorts      : {}
```


Or maybe this is not the correct way to go?

**Reply by NikolajBjorner** [reply]
I was unable to rebuild your solution directly - my powershell skills are too poor to be able to work around the error messages i got. You can save the state of the Optimize context into an SMTLIB string by using the ToString() method, such as
Console.WriteLine($"{s}"); should print 's' into a text format that is easy to exchange and add to this thread.

**Reply by NikolajBjorner** [reply]
> Does it mean that arrays in Z3 is not for, kind of, "storing" some data and iterating through it, but rather for something different?

Right, I wouldn't conflate "arrays" used in minizink with "arrays" used in SMT solvers. Same name, but not the same target use.

**Reply by eosfor** [reply]
Thanks for all your help @NikolajBjorner , I added the quick and dirty piece that exports the model into a file. The file turns out to be big, so I pushed it to the repo [here](https://raw.githubusercontent.com/eosfor/z3testps/master/smtlib-model.txt)

**Reply by eosfor** [reply]
> I was unable to rebuild your solution directly - my powershell skills are too poor to be able to work around the error messages i got. You can save the state of the Optimize context into an SMTLIB string by using the ToString() method, such as Console.WriteLine($"{s}"); should print 's' into a text format that is easy to exchange and add to this thread.

there is a ps1 file in the solution, with examples and comments, if it can help. 

Just to summarize, you build it by running this in the root of the repo

```console
dotnet build && dotnet publish --self-contained --use-current-runtime -c Debug -o c:\temp\z3testps\
```

then `cd` into the output folder, in my case it is `c:\temp\z3testps\`, import the module dll, and run the command, supplying necessary `source` and `target` data

```pwsh
cd "c:\temp\z3testps\"
Import-Module .\z3testps.dll -Verbose
$pid | clip ## this is to put the pwsh process id into the clipboard, so you can then pasete it to the VS to be able to attach to the process

$sourceVMs = Import-Csv .\vmdata.csv
$sourceVMs  | % { $_.cpu = [int]$_.cpu; $_.ram = [int]$_.ram; $_.datadisk = [int]$_.datadisk; }

$targetSizes = import-csv ".\vmCostACUData.csv"

$m = Start-Z3ModelCalculation -SourceVM $sourceVMs -TargetVM $targetSize
```

the variable `$m` at the end is an array, containing a few elements: the model, two optimization objectives, and the context. I do it so I can  explore them from the command line

```csharp
            if (s.Check() == Status.SATISFIABLE)
            {
                var m = s.Model;
                WriteObject(m);
                WriteObject(totalAcuHandle);
                WriteObject(totalPriceHandle);
                WriteObject(ctx);
            }
```

![image](https://user-images.githubusercontent.com/9363027/151714167-56935d6d-7ada-48fa-975c-555684161f56.png)

**Reply by NikolajBjorner** [reply]
You seem to be using "integer" variables for z3 where you use 0-1 variables in the other solver.
The problem is unbounded for integers. Maybe you intend to solve a Pseudo-Boolean domain?
You can use Booleans and PbEq for Boolean equalities. You can also declare all variables as integers and add bounds 0 <= x <= 1 for each x and keep using integer expressions. The format for PbEq is perhaps arcane, but captures the intent.
You can also use IfThenElse expressions to get from Boolean variables to integer sums.
(+ (if x 1 0) (if y 2 0) ... ) etc.

**Reply by eosfor** [reply]
I'll try. The only question I have about PbEq is what is the first parameter? As I see [here](https://stackoverflow.com/questions/43081929/k-out-of-n-constraint-in-z3py), `args` is my array, `k` is the number of "ones" I need. and what is the `coeffs`?

```csharp
public BoolExpr MkPBEq(int[] coeffs, BoolExpr[] args, int k)
```

**Reply by NikolajBjorner** [reply]
In your case you have equalities of the form (+ x y z .. ) = 1
the coefficients to x, y, z are all 1.

MkPBEq(all_ones, variables, 1);

I converted your example file to work with Booleans. z3 is **not** able to efficiently solve it. It spends most time detecting that you have a VERY large number of binary cores (the mutex constraint is that two variables that cannot be 1 at the same time). This is evident from the constraints, of course, but not something the solver is tuned for at all. 

It does finish solving the system after several minutes, but most time seems spent on simple inferences that could be used
much more efficiently if z3 had been trained on such shapes of constraints and possibly the core guided solver isn't the best for this shape.

There is limited detection for mutex but it doesn't find these and even then it is a relatively weak heuristic.

**Reply by eosfor** [reply]
I'm trying to do it too. I also want to study how to extract the results from the model, similar to what I do with or-tools. Maybe you can paste the relevant pieces here :D?

Actually, I planned to convert all this C# code into pure PowerShell and try it in [.NET interactive notebooks](https://docs.microsoft.com/en-us/shows/vs-code-livestreams/net-interactive-notebooks-with-csharp-fsharp). The aim here was to open the modeling capability for those, who are not familiar with C# and are more oriented at scripting. It makes it available right in the command line. However the complexity of the C# interface itself, kind of, still makes it hard to use.

Does it make sense to "train" the solver to be able to deal with it?

BTW, separate question, does it make sense to interface Z3 to minizinc, as or-tools does, for example? Minizinc seems to provide a pretty convenient language for modeling. I know there is smt-lib, but it does not look as simple. Plus minizinc has a set of pretty good courses on Coursera, and that will give another push for Z3, perhaps. Just my IMHO.

Thanks again, @NikolajBjorner! All this is really cool and I still have much to learn. I wish there was a course on the topic, covering the concepts, C# API with examples, like [here](https://www.coursera.org/learn/basic-modeling),

**Reply by eosfor** [reply]
I did it!

I did not have a chance to verify all the stuff yet, but as far as I can see the optimization objective values look the same as or-tools return. Now the question is, can Z3 do it faster. I wish it could

**costs minimum**
> OR-TOOLS
CpSolverResponse summary:
status: OPTIMAL
**objective: 20010**
best_bound: 20010
booleans: 0
conflicts: 0
branches: 0
propagations: 0
integer_propagations: 1
restarts: 0
lp_iterations: 0
walltime: 0.636374
usertime: 0.636374
deterministic_time: 0.00416944
gap_integral: 0

> Z3
Lower         : 20010
Upper         : 20010
**Value         : 20010**
LowerAsVector : {0, 20010, 0}
UpperAsVector : {0, 20010, 0}

**performance maximum**
> OR-TOOLS
CpSolverResponse summary:
status: OPTIMAL
**objective: 584240**
best_bound: 584240
booleans: 260
conflicts: 1095
branches: 2340
propagations: 88449
integer_propagations: 18152
restarts: 507
lp_iterations: 0
walltime: 6.82701
usertime: 6.82701
deterministic_time: 0.422987
gap_integral: 0.00604726

> Z3
Lower         : 584240
Upper         : 584240
**Value         : 584240**
LowerAsVector : {0, 584240, 0}
UpperAsVector : {0, 584240, 0}


Thank you SO MUCH, @NikolajBjorner!

🥳 🎉

**Reply by NikolajBjorner** [reply]
It solves in about 389 seconds, which is some orders of magnitude slower than what should be possible. 
Small point fixes will not help and algorithmic changes are required to solve this kind of problems.
I learn from all examples; and learned a few minor issues that led to minor improvements, lurking under the covers.

I am not familiar with minizink to smtlib conversion efforts. There is https://link.springer.com/chapter/10.1007/978-3-030-58942-4_10, which targets the OMT solver. It is maybe good to expose z3 to the OR problem domains, which it will not be tuned for, but likely a time zink, I mean sink, given that the main backends to minizink are tuned for the problem sets already formulated for it while z3 is not at all.

**Reply by NikolajBjorner** [reply]
information like this:

booleans: 0
conflicts: 0
branches: 0
propagations: 0
integer_propagations: 1
restarts: 0
lp_iterations: 0

suggests the first objective is solved by pre-processing.
I guess the optimistic bound is feasible and therefore detected at once.

**Reply by eosfor** [reply]
In the end, I came up with this:

the PowerShell module to try Z3 and OR-TOOLS for this task: https://github.com/eosfor/AzureVmCalc. Z3 is hidden so far, due to the slowness. @NikolajBjorner doe it makes sense to register this as an, say, `improvement request` or something?

The [.NET interactive notebook with the or-tools example](https://github.com/eosfor/scripting-notes/blob/main/notebooks/en/vmOptimization-or-tools.ipynb). The Z3 one should look the same. 

Thanks, @NikolajBjorner for all your help!
